File: viz/src/geometry.ts
/**
 * Road frame geometry, mirroring `scenario_runner.model.geometry`.
 *
 * Kept deliberately small and duplicated rather than shipped from Python,
 * because the alternative is a build step that couples the browser bundle to
 * the Python package. The two implementations are three lines of trigonometry
 * each and are checked against one another by eye on the rendered lane
 * boundaries: a mismatch is immediately visible as a road that does not follow
 * the vehicles.
 */

import type { Road, Shape } from "./types.js";

export interface Point {
  readonly x: number;
  readonly y: number;
}

/** Cartesian position of the road point at arc length `s` and lateral offset `d`. */
export function frenetToCartesian(road: Road, s: number, d: number): Point {
  if (road.curvature === 0) {
    return { x: s, y: d };
  }
  const radius = 1 / road.curvature;
  const theta = s * road.curvature;
  const effective = radius - d;
  return {
    x: effective * Math.sin(theta),
    y: radius - effective * Math.cos(theta),
  };
}

/** Lateral offset of the centre of a lane. Lane 0 is the rightmost lane. */
export function laneOffset(road: Road, lane: number): number {
  return lane * road.lane_width;
}

/** Polyline for one lane boundary, sampled along the road. */
export function laneBoundary(road: Road, boundary: number, samples = 240): Point[] {
  const offset = laneOffset(road, boundary) - road.lane_width / 2;
  const points: Point[] = [];
  for (let index = 0; index <= samples; index += 1) {
    const s = (road.length * index) / samples;
    points.push(frenetToCartesian(road, s, offset));
  }
  return points;
}

/** The four corners of a vehicle footprint in world coordinates. */
export function footprint(x: number, y: number, yaw: number, shape: Shape): Point[] {
  const cos = Math.cos(yaw);
  const sin = Math.sin(yaw);
  const half = shape.width / 2;
  const local: [number, number][] = [
    [shape.rear_offset, -half],
    [shape.front_offset, -half],
    [shape.front_offset, half],
    [shape.rear_offset, half],
  ];
  return local.map(([along, across]) => ({
    x: x + along * cos - across * sin,
    y: y + along * sin + across * cos,
  }));
}

File: src/scenario_runner/model/geometry.py
"""Road frame conversions for a straight or constant curvature road.

Coordinates follow the usual road convention: ``s`` is arc length along the
lane 0 reference line, ``d`` is lateral offset with positive to the left, and
the Cartesian frame has the reference line starting at the origin heading along
positive ``x``.

For curvature ``k = 1 / R`` (positive turning left, centre of curvature at
``(0, R)``) a point at ``(s, d)`` is::

    theta = s * k
    u     = R - d
    x     = u * sin(theta)
    y     = R - u * cos(theta)

which reduces to ``(s, d)`` as ``k`` tends to zero.
"""

from __future__ import annotations

import math

from scenario_runner.model.scenario import RoadSpec

__all__ = ["cartesian_to_frenet", "frenet_to_cartesian", "heading_at"]


def frenet_to_cartesian(road: RoadSpec, s: float, d: float) -> tuple[float, float, float]:
    """Return ``(x, y, heading)`` for the road point at arc length ``s``, offset ``d``."""
    curvature = road.curvature
    if curvature == 0.0:
        return s, d, 0.0
    radius = 1.0 / curvature
    theta = s * curvature
    effective = radius - d
    x = effective * math.sin(theta)
    y = radius - effective * math.cos(theta)
    return x, y, theta


def cartesian_to_frenet(road: RoadSpec, x: float, y: float) -> tuple[float, float]:
    """Return ``(s, d)`` for a Cartesian point. Exact inverse of ``frenet_to_cartesian``."""
    curvature = road.curvature
    if curvature == 0.0:
        return x, y
    radius = 1.0 / curvature
    sign = math.copysign(1.0, radius)
    distance = math.hypot(x, radius - y)
    theta = math.atan2(sign * x, sign * (radius - y))
    return theta * radius, radius - sign * distance


def heading_at(road: RoadSpec, s: float) -> float:
    """Heading of the road reference line at arc length ``s``."""
    return s * road.curvature
